src/flint/types/fmpq_mpoly.pyx

```diff
@@ -0,0 +1,1030 @@
+
+    def subs(self, dict_args) -> fmpq_mpoly:
+        """
+        Partial evaluate this polynomial with select constants. Keys must be generator names,
+        all values must be fmpq.
+
+            >>> from flint import Ordering
+            >>> ctx = fmpq_mpoly_ctx.get_context(2, Ordering.lex, 'x')
+            >>> f = ctx.from_dict({(0, 0): 1, (1, 0): 2, (0, 1): 3, (1, 1): 4})
+            >>> f.subs({"x1": 0})
+            2*x0 + 1
+
+        """
+        cdef:
+            fmpq_mpoly res
+            slong i, nargs
+
+        args = tuple((i, dict_args[x]) for i, x in enumerate(self.ctx.names()) if x in dict_args)
+        nargs = len(args)
+
+        # If we've been provided with an invalid generator name then the length of our filter
+        # args will be less than what we've been provided with.
+        if nargs < len(dict_args):
+            raise ValueError("unknown generator name provided")
+
+        args_fmpq = tuple(any_as_fmpq(v) for _, v in args)
+        for arg in args_fmpq:
+            if arg is NotImplemented:
+                raise TypeError(f"cannot coerce argument ('{arg}') to fmpq")
+
+        # Partial application with args in Q. We evaluate the polynomial one variable at a time
+        res = create_fmpq_mpoly(self.ctx)
+
+        fmpq_mpoly_set(res.val, self.val, self.ctx.val)
+        for (i, _), arg in zip(args, args_fmpq):
+            if fmpq_mpoly_evaluate_one_fmpq(res.val, res.val, i, (<fmpq>arg).val, self.ctx.val) == 0:
+                raise ValueError("unreasonably large polynomial")  # pragma: no cover
+        return res
+
+    def compose(self, *args) -> fmpq_mpoly:
+        """
+        Compose this polynomial with other fmpq_mpolys. All arguments must share the same context, it may different
+        from this polynomials context.
+
+            >>> from flint import Ordering
+            >>> ctx = fmpq_mpoly_ctx.get_context(1, Ordering.lex, 'x')
+            >>> ctx1 = fmpq_mpoly_ctx.get_context(2, Ordering.lex, 'y')
+            >>> f = ctx.from_dict({(2,): 1})
+            >>> g = ctx1.from_dict({(1, 0): 1, (0, 1): 1})
+            >>> f
+            x^2
+            >>> g
+            y0 + y1
+            >>> f.compose(g)
+            y0^2 + 2*y0*y1 + y1^2
+
+        """
+        cdef:
+            fmpq_mpoly res
+            fmpq_mpoly_ctx res_ctx
+            fmpq_mpoly_vec C
+            slong i, nvars = self.ctx.nvars(), nargs = len(args)
+
+        if nargs < nvars:
+            raise ValueError("not enough arguments provided")
+        elif nargs > nvars:
+            raise ValueError("more arguments provided than variables")
+        elif not all(typecheck(arg, fmpq_mpoly) for arg in args):
+            raise TypeError("all arguments must be fmpq_mpolys")
+
+        res_ctx = (<fmpq_mpoly> args[0]).ctx
+        if not all((<fmpq_mpoly> args[i]).ctx is res_ctx for i in range(1, len(args))):
+            raise IncompatibleContextError("all arguments must share the same context")
```

Comment:
The base case is a polynomial ring with no generators:
```python
In [2]: ctx = fmpq_mpoly_ctx(0, Ordering.lex, [])

In [3]: p = fmpq_mpoly('2', ctx)

In [4]: p.compose()
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 p.compose()

File fmpq_mpoly.pyx:683, in flint.types.fmpq_mpoly.fmpq_mpoly.compose()

IndexError: tuple index out of range
```
When there are no variables there can be no arguments but we need to know what context to use for the output so there should probably be a way to specify that like:
```python
p.compose(ctx=ctx2)
```

Reply:
This should be fixed now, a context can be provided, if one isn't, then it is inferred from the arguments. Thanks.